File: tests/core/transducer.cpp
// File: transducer.cpp
// Project: core
// Created Date: 02/12/2022
// Author: Shun Suzuki
// -----
// Last Modified: 21/12/2022
// Modified By: Shun Suzuki (suzuki@hapis.k.u-tokyo.ac.jp)
// -----
// Copyright (c) 2022 Shun Suzuki. All rights reserved.
//

#if _MSC_VER
#pragma warning(push)
#pragma warning(disable : 26439 26495 26812)
#endif
#include <gtest/gtest.h>
#if _MSC_VER
#pragma warning(pop)
#endif

#include <autd3/core/transducer.hpp>

#include "test_utils.hpp"

using autd3::core::Quaternion;
using autd3::core::Vector3;
using autd3::driver::autd3_float_t;

TEST(CoreTransducer, Transducer) {
  const auto rot = Eigen::AngleAxis(autd3::driver::pi / autd3_float_t{2}, Vector3::UnitZ()) * Eigen::AngleAxis(autd3_float_t{0}, Vector3::UnitY()) *
                   Eigen::AngleAxis(autd3_float_t{0}, Vector3::UnitX());

  autd3::core::Transducer tr(1, Vector3(10, 20, 30), rot);

  ASSERT_NEAR_VECTOR3(tr.position(), Vector3(10, 20, 30), 1e-3);
  ASSERT_NEAR_VECTOR3(tr.x_direction(), Vector3(0, 1, 0), 1e-3);
  ASSERT_NEAR_VECTOR3(tr.y_direction(), Vector3(-1, 0, 0), 1e-3);
  ASSERT_NEAR_VECTOR3(tr.z_direction(), Vector3(0, 0, 1), 1e-3);

  ASSERT_EQ(tr.id(), 1);

  tr.set_cycle(3000);
  ASSERT_EQ(tr.cycle(), 3000);
  tr.set_frequency(70e3);
  ASSERT_NEAR(tr.frequency(), 70e3, 15.0);

  ASSERT_NEAR(tr.wavelength(), 4.857142857142857142857142857L, 1e-3);
  ASSERT_NEAR(tr.wavenumber(), 1.293596975007561871293279075L, 1e-3);
}

File: src/autd3-core/src/geometry/transducer.rs
/*
 * File: transducer.rs
 * Project: geometry
 * Created Date: 04/05/2022
 * Author: Shun Suzuki
 * -----
 * Last Modified: 18/07/2023
 * Modified By: Shun Suzuki (suzuki@hapis.k.u-tokyo.ac.jp)
 * -----
 * Copyright (c) 2022-2023 Shun Suzuki. All rights reserved.
 *
 */

use autd3_driver::{float, PI};

use super::{Quaternion, UnitQuaternion, Vector3};

fn get_direction(dir: Vector3, rotation: &UnitQuaternion) -> Vector3 {
    let dir: UnitQuaternion = UnitQuaternion::from_quaternion(Quaternion::from_imag(dir));
    (rotation * dir * rotation.conjugate()).imag().normalize()
}

pub trait Transducer: Send + Sync {
    type Gain: autd3_driver::operation::GainOp;
    type Sync: autd3_driver::operation::SyncOp;
    type GainSTM: autd3_driver::operation::GainSTMOp;

    /// Create transducer
    fn new(id: usize, pos: Vector3, rot: UnitQuaternion) -> Self;
    /// Affine transformation
    fn affine(&mut self, pos: Vector3, rot: UnitQuaternion);
    /// Calculate the phase of the transducer to align the phase at the specified position
    fn align_phase_at(&self, pos: Vector3, sound_speed: float) -> float {
        (pos - self.position()).norm() * self.wavenumber(sound_speed)
    }
    /// Get the position of the transducer
    fn position(&self) -> &Vector3;
    /// Get the rotation of the transducer
    fn rotation(&self) -> &UnitQuaternion;
    /// Get the index of the transducer
    fn idx(&self) -> usize;
    /// Get the x-direction of the transducer
    fn x_direction(&self) -> Vector3 {
        get_direction(Vector3::x(), self.rotation())
    }
    /// Get the y-direction of the transducer
    fn y_direction(&self) -> Vector3 {
        get_direction(Vector3::y(), self.rotation())
    }
    /// Get the z-direction (axial direction) of the transducer
    fn z_direction(&self) -> Vector3 {
        get_direction(Vector3::z(), self.rotation())
    }
    /// Get the frequency of the transducer
    fn frequency(&self) -> float;
    /// Get the modulation delay of the transducer
    fn mod_delay(&self) -> u16;
    /// Set the modulation delay of the transducer
    fn set_mod_delay(&mut self, value: u16);
    /// Get the wavelength of the transducer
    fn wavelength(&self, sound_speed: float) -> float {
        sound_speed / self.frequency()
    }
    /// Get the wavenumber of the transducer
    fn wavenumber(&self, sound_speed: float) -> float {
        2.0 * PI * self.frequency() / sound_speed
    }
    /// Get the cycle of the transducer
    fn cycle(&self) -> u16;
}
